platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,482 @@
+                    throw z;
+                } finally {
+                    snapshotCompactionLock.release();
+                }
+            });
+            allDataItemsProcessed = true;
+        } finally {
+            // Even if the thread is interrupted, make sure the new compacted file is properly closed
+            // and is included to future compactions
+            snapshotCompactionLock.acquire();
+            try {
+                // Finish writing the last file. In rare cases, it may be an empty file
+                finishCurrentCompactionFile();
+                // Clear compaction start time
+                currentCompactionStartTime.set(null);
+                if (allDataItemsProcessed) {
+                    // Close the readers and delete compacted files
+                    dataFileCollection.deleteFiles(filesToCompact);
+                }
+            } finally {
+                snapshotCompactionLock.release();
+            }
+        }
+
+        return newCompactedFiles;
+    }
+
+    // visible for testing
+    int getMinNumberOfFilesToCompact() {
+        return config.minNumberOfFilesInCompaction();
+    }
+
+    /**
+     * Opens a new file for writing during compaction. This method is called, when compaction is
+     * started. If compaction is interrupted and resumed by data source snapshot using {@link
+     * #pauseCompaction()} and {@link #resumeCompaction()}, a new file is created for writing using
+     * this method before compaction is resumed.
+     * <p>
+     * This method must be called under snapshot/compaction lock.
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    private void startNewCompactionFile(int compactionLevel) throws IOException {
+        final Instant startTime = currentCompactionStartTime.get();
+        assert startTime != null;
+        final DataFileWriter<?> newFileWriter = dataFileCollection.newDataFile(startTime, compactionLevel);
+        currentWriter.set(newFileWriter);
+        final Path newFileCreated = newFileWriter.getPath();
+        newCompactedFiles.add(newFileCreated);
+        final DataFileMetadata newFileMetadata = newFileWriter.getMetadata();
+        final DataFileReader<?> newFileReader =
+                dataFileCollection.addNewDataFileReader(newFileCreated, newFileMetadata);
+        currentReader.set(newFileReader);
+    }
+
+    /**
+     * Closes the current compaction file. This method is called in the end of compaction process,
+     * and also before a snapshot is taken to make sure the current file is fully written and safe
+     * to include to snapshots.
+     * <p>
+     * This method must be called under snapshot/compaction lock.
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    private void finishCurrentCompactionFile() throws IOException {
+        currentWriter.get().finishWriting();
+        currentWriter.set(null);
```

Comment:
Before we null these references, do we need to close the writer and reader?

Reply:
`currentWriter.get().finishWriting()` is closing the file channel but your observation is totally correct about the reader.